android/src/main/java/io/grpc/android/UdsSocket.java

```diff
@@ -0,0 +1,304 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.android;
+
+import android.net.LocalSocket;
+import android.net.LocalSocketAddress;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.io.FilterInputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.SocketChannel;
+
+@SuppressWarnings("UnsynchronizedOverridesSynchronized") // Rely on LocalSocket's synchronization
+class UdsSocket extends Socket {
+
+  private final LocalSocket localSocket;
+  private final LocalSocketAddress localSocketAddress;
+
+  @GuardedBy("this")
+  private boolean closed = false;
+
+  @GuardedBy("this")
+  private boolean inputShutdown = false;
+
+  @GuardedBy("this")
+  private boolean outputShutdown = false;
+
+  public UdsSocket(LocalSocketAddress localSocketAddress) {
+    this.localSocketAddress = localSocketAddress;
+    localSocket = new LocalSocket();
+  }
+
+  @Override
+  public void bind(SocketAddress bindpoint) {
+    // no-op
+  }
+
+  @Override
+  public synchronized void close() throws IOException {
+    if (closed) {
+      return;
+    }
+    if (!inputShutdown) {
+      shutdownInput();
+    }
+    if (!outputShutdown) {
+      shutdownOutput();
+    }
+    localSocket.close();
+    closed = true;
+  }
+
+  @Override
+  public void connect(SocketAddress endpoint) throws IOException {
+    localSocket.connect(localSocketAddress);
+  }
+
+  @Override
+  public void connect(SocketAddress endpoint, int timeout) throws IOException {
+    localSocket.connect(localSocketAddress, timeout);
+  }
+
+  @Override
+  public SocketChannel getChannel() {
+    throw new UnsupportedOperationException("getChannel() not supported");
+  }
+
+  @Override
+  public InetAddress getInetAddress() {
+    throw new UnsupportedOperationException("getInetAddress() not supported");
+  }
+
+  @Override
+  public InputStream getInputStream() throws IOException {
+    return new FilterInputStream(localSocket.getInputStream()) {
+      @Override
+      public void close() throws IOException {
+        UdsSocket.this.close();
+      }
+    };
+  }
+
+  @Override
+  public boolean getKeepAlive() {
+    throw new UnsupportedOperationException("Unsupported operation getKeepAlive()");
```

Comment:
Would we want to explain in comment(s) why some of the `Socket` operations are not supported?